src/Data/Nat/Binary/Base.agda

```diff
@@ -112,10 +117,29 @@ toℕ zero     =  0
 toℕ 2[1+ x ] =  2 ℕ.* (ℕ.suc (toℕ x))
 toℕ 1+[2 x ] =  ℕ.suc (2 ℕ.* (toℕ x))
 
--- Costs O(n), could be improved using `_/_` and `_%_`
 fromℕ : ℕ → ℕᵇ
-fromℕ 0         = zero
-fromℕ (ℕ.suc n) = suc (fromℕ n)
+fromℕ = <-rec (λ _ → ℕᵇ) λ
+  { ℕ.zero _ → zero
+  ; (ℕ.suc n) p → case n divMod 2 of λ
+    { (result q Fin.zero prop) →
+      -- n is even so suc n is odd
+      1+[2 p q $ ℕ.s≤s $ begin
+        q       ≡˘⟨ *-identityʳ q ⟩
+        q ℕ.* 1 ≤⟨ *-monoʳ-≤ q (ℕ.s≤s ℕ.z≤n) ⟩
+        q ℕ.* 2 ≡˘⟨ prop ⟩
+        n       ∎
+      ]
+    ; (result q (Fin.suc Fin.zero) prop) →
+      -- n is odd so suc n is even
+      2[1+ p q $ ℕ.s≤s $ begin
+        q       ≡˘⟨ *-identityʳ q ⟩
+        q ℕ.* 1 ≤⟨ *-monoʳ-≤ q (ℕ.s≤s ℕ.z≤n) ⟩
+        q ℕ.* 2 <⟨ ≤-reflexive (≡.sym prop) ⟩
```

Comment:
I'm not happy with this line. Is there a cleaner way?

Reply:
I think it can be written as `≡˘⟨ prop ⟩`

Reply:
Sadly not. That gives:

```agda
Goal: n ≡ q ℕ.* 2
Have: n ≡ ℕ.suc (q ℕ.* 2)
```

What I have is to convince Agda that the extra `suc` is OK to have in an inequality proof. Note I'm using `≤-reflexive` and then using it as a proof of `_<_`.

Reply:
Well in that case, I'd split it into two lines:
```
q ℕ.* 2        <⟨ n<1+n _ ⟩
suc (q ℕ.* 2)  ≡˘⟨ prop ⟩
n              ∎
```